src/frequenz/channels/util/_select.py

```diff
@@ -8,199 +8,223 @@
+def is_selected(
+    selected: Selected[Any], receiver: _base_classes.Receiver[_T]
+) -> TypeGuard[Selected[_T]]:
+    """Check if the given receiver was selected.
+
+    This function is used in conjuction with the `Selected` class to determine which
+    receiver was selected in a select operation.
+
+    It also works as a type guard to narrow the type of the `Selected` instance to the
+    type of the receiver.
+
+    Args:
+        selected: The result of a select operation.
+        receiver: The receiver to check.
+
+    Returns:
+        True if the given receiver was selected, False otherwise.
+    """
+    if handled := selected._recv is receiver:
+        selected._handled = True
+    return handled
+
+
+async def select(
+    *receivers: _base_classes.Receiver[Any], timeout: _datetime.timedelta | None = None
+) -> AsyncIterator[Selected[Any]]:
+    """Wait until there is a message in any of the receivers.
+
+    Returns `True` if there is a message available, and `False` if all
+    receivers have closed.
+
+    Yields:
+        A set with the receivers that are ready to be consumed.
+
+    Raises:
+        BaseException: if the receivers raise any exceptions.
+
+    # noqa: DAR401 exc (https://github.com/terrencepreilly/darglint/issues/181)
+    """
+    receivers_map: dict[str, _base_classes.Receiver[Any]] = {
+        str(hash(r)): r for r in receivers
+    }
+    pending: set[_asyncio.Task[bool]] = set()
+
+    async def _recv_task(recv: _base_classes.Receiver[Any]) -> bool:
+        try:
+            return await recv.ready()
+        except _asyncio.CancelledError:
+            return False
+
+    try:
+        for name, recv in receivers_map.items():
+            pending.add(_asyncio.create_task(_recv_task(recv), name=name))
+
+        while pending:
+            done, pending = await _asyncio.wait(
+                pending,
+                return_when=_asyncio.FIRST_COMPLETED,
+                timeout=timeout.total_seconds() if timeout else None,
+            )
+            ready_set: set[_base_classes.Receiver[Any]] = set()
+            for task in done:
+                name = task.get_name()
+                recv = receivers_map[name]
+                # This will raise if there was an exception in the task, as this is
+                # unexpected (Receiver.ready() should not raise).
+                exc = task.exception()
+                if exc is not None:
+                    raise exc
+                ready_set.add(recv)
+
+            for ready in ready_set:
+                selected = Selected(ready)
+                yield selected
+                assert selected._handled, f"{selected} was not handled"
+
+            for task in done:
+                receiver_active = task.result()
+                if not receiver_active:
+                    continue
+                name = task.get_name()
+                recv = receivers_map[name]
+                pending.add(_asyncio.create_task(_recv_task(recv), name=name))
```

Comment:
It might be a leftover from a previous implementation.  At some point there was a reason for it, but maybe that reason doesn't exist any more.  I'll try to make all in one loop and if I can't I will comment on the why :+1:

Reply:
Removed the extra loops :rocket: